### PR TITLE
This PR  alters the DQ_VrepInterface() properties from 'private' to 'protected'

### PR DESCRIPTION
**Pull request by ffasilva**
# Main instructions

By submitting this pull request, you automatically agree that you have read and accepted the following conditions:

- Anyone wanting to propose a new modification or introduce new functionality should reach out to the team first, as proposed modifications that do not comply with the library's development philosophy and style, do not follow the library's architecture, do not introduce a clear and general benefit to the library other than to the person who proposed the modification will likely be rejected with no further discussion.
- Support for DQ Robotics is given voluntarily and it is not the developers' role to educate and/or convince anyone of their vision.
- Any possible response and its timeliness will be based on the relevance, accuracy, and politeness of a request and the following discussion.
- You are familiar with the [development workflow](https://github.com/dqrobotics/matlab/blob/master/CONTRIBUTING.md).
- Each pull request should contain only individual changes (several changes of the same type **are allowed** on the same pull request).
- Refactoring or modifying internal implementation that is working is not allowed unless comprehensively discussed with and approved by @bvadorno and @mmmarinho.

## Description of changes
This PR  alters the DQ_VrepInterface() properties from `private` to `protected`.

## Rationale
The DQ_VrepInterface() properties being `private` prevent inheritance as the derived classes cannot get access to them. This pull request fix this problem, and keep the properties protected from modifications without `set` methods, by altering their scope to `protected`.

### CODE REVIEW

**Comment by mmmarinho** [comment]
@dqrobotics/developers

I’m not sure if there was a discussion about this before, but I’d be careful making everything protected as a given.

In general, making something protected in a library means that we, the developers, are making an API contract with our users that they can inherit and use those attributes/methods and we will not change those without a prior deprecation step.

Whereas private attributes/methods mean that we make no such contract and we are free to change the internal implementation, without worries about whose code my break with those changes.

I’m just trying to think ahead here. For instance, there are methods in the CPP version I’d recommend against making protected.

Alternatively, we can make it explicit somewhere that we do not make any promises regarding the protected methods. But I’d say that’s unusual.

**Comment by bvadorno** [comment]
> @dqrobotics/developers
> 
> I’m not sure if there was a discussion about this before, but I’d be careful making everything protected as a given.
> 
> In general, making something protected in a library means that we, the developers, are making an API contract with our users that they can inherit and use those attributes/methods and we will not change those without a prior deprecation step.
> 
> Whereas private attributes/methods mean that we make no such contract and we are free to change the internal implementation, without worries about whose code my break with those changes.
> 
> I’m just trying to think ahead here. For instance, there are methods in the CPP version I’d recommend against making protected.
> 
> Alternatively, we can make it explicit somewhere that we do not make any promises regarding the protected methods. But I’d say that’s unusual.

Hi @mmmarinho,

@ffasilva and I discussed it offline. You make a good point, but the idea is not to change every private statement in the DQ Robotics library to `protected`.  Instead, the rationale behind modifying some properties of the V-REP interface to protected is that the interface is (understandably) rather incomplete. Therefore, if one needs to extend it to include new methods for their applications, they can't do it through inheritance if crucial properties, such as `vrep` and `client_id` are private. The alternative would be to create another class, instantiate a `DQ_VrepInterface` object, and then get those pieces of information using `set` and `get` methods, which, depending on what needs to be done, might be very clumsy. Moreover, if later on, other crucial functionalities are implemented, such as destructors, robust error handling, etc., non-inherited classes will not benefit from those additions.

This is more robust from the user's point of view, _assuming the tacit contract that we won't be changing methods without using our deprecation policy_, as you correctly pointed out.

I suggest changing those properties to private while the V-Rep interface is being improved, such that any improvements one makes might coexist with and benefit from those improvements. What do you think?

Best wishes,
Bruno

**Comment by mmmarinho** [comment]
> Hi @mmmarinho,
> 
> @ffasilva and I discussed it offline. You make a good point, but the idea is not to change every private statement in the DQ Robotics library to `protected`. Instead, the rationale behind modifying some properties of the V-REP interface to protected is that the interface is (understandably) rather incomplete. Therefore, if one needs to extend it to include new methods for their applications, they can't do it through inheritance if crucial properties, such as `vrep` and `client_id` are private. The alternative would be to create another class, instantiate a `DQ_VrepInterface` object, and then get those pieces of information using `set` and `get` methods, which, depending on what needs to be done, might be very clumsy. Moreover, if later on, other crucial functionalities are implemented, such as destructors, robust error handling, etc., non-inherited classes will not benefit from those additions.
> 
> This is more robust from the user's point of view, _assuming the tacit contract that we won't be changing methods without using our deprecation policy_, as you correctly pointed out.
> 
> I suggest changing those properties to private while the V-Rep interface is being improved, such that any improvements one makes might coexist with and benefit from those improvements. What do you think?
> 
> Best wishes, Bruno

Thanks, @bvadorno 

What you said is perfectly aligned with what I meant.

Concretely, making `clientID` and `vrep` _protected_ makes a lot of sense. In fact, just to add to your comment, currently, it would not only be clumsy to extend `DQ_VrepInterface`, but it would also be pretty much impossible because those are not exposed at all. 

For a less clumsy extension, `handles_map` should also be _protected_ (this PR is changing it to _protected_ already, which is why I'm mentioning this). However, changing `handles_map` to _protected_ needs us to make `handle_from_string_or_handle()` and `element_from_string()` also _protected_ for `handles_map` to be of any use. That implementation has been robust for years, so I wouldn't be against changing it, but there is an [issue](https://github.com/dqrobotics/cpp-interface-vrep/issues/9#issue-1404564992) that will require us to change the map implementation a bit. If there isn't a use case with a certain priority that you are envisioning right now, it might be reasonable to postpone the change of `handles_map`'s access to sometime after this release.

----------------
**TL;DR** My suggestion would be to make _protected_

```matlab
clientID
vrep
```

or, depending on the trade-off,

```matlab
clientID
vrep
handles_map
element_from_string()
handle_from_string_or_handle()
```
----------------

Do you think this makes sense?

Kind regards,
Murilo

_Extra comment#1: For the future developer that will implement this on CPP, note that the attribute `vrep` does not exist in CPP, but that has nothing to do with dqrobotics and only to do with CoppeliaSim, so we don't have to worry about compatibility in that case._

_Extra comment#2: In CPP I wouldn't change anything else besides what was mentioned here to _protected_. The discussion becomes considerably more complicated than for the MATLAB implementation._

**Comment by bvadorno** [comment]
Hi @mmmarinho,

That makes perfect sense! @ffasilva, for the application we were discussing offline, I think that only 
```
clientID
vrep
```
must be protected. Can you confirm that? If yes, let's proceed with Murilo's first suggestion.

Best wishes,
Bruno

**Comment by ffasilva** [comment]
Hi @mmmarinho,

Although I discussed the modifications with @bvadorno before the pull request, I hadn't fully understood the motivations behind the private properties in the class. Thanks for the clarification!

Your suggestions make sense. For immediate use, we only need to have `clientID` and `vrep` set as `protected`. This already allows me to make the necessary class extensions through inheritance.

Kind regards,
Frederico

**Review by bvadorno — APPROVED** [review]

**Review by bvadorno — APPROVED** [review]
